### PR TITLE
[CDU] Fuel Predictions and Functional Init B

### LINKED ISSUE #219: Not all elements on Int B MCDU page functioning
Trip time, trip fule not calculated from flight plan, 
altn time and fuel not functional

https://www.youtube.com/watch?v=c7UJlJMJdfk

### LINKED ISSUE #689: [REQUEST] Fuel Predictions aren't calculated
**Is your feature request related to a problem? Please describe.**
Yes, the current FUEL PRED page remains blank of any information. Fuel prediction is essential to any pilot to ensure they have enough fuel onboard for their trip as well as seeing fuel information during flight. It is also used after an engine is started and INIT B is no longer used.

**Describe the feature you'd like to see implemented**
The fuel predictions page implemented per the fcom 

**References**

FCOM 4.03.20 P14

**Additional context**

http://www.a320dp.com/A320_DP/nav-flight-management/sys-14.10.0.html

Discord username: Lucky38#3550

### LINKED ISSUE #726: [BUG] ZFW CG Incorrect
**Mod Version**
Master

**Describe the bug**
On INIT page 2 of the FMGC, the ZFWCG appears to now be automatically entered. However, the value entered appears to be the value taken directly from the flight planner weights page. This value is not the zero fuel weight (ZFW) CG, it is the CG for the current weight. You can verify this by moving the fuel slider to zero, so that the airplane weight is the zero fuel weight. The CG changes. It is this CG that is the ZFW CG.

**To Reproduce**
see above

1.
2.
3.

**Expected behavior**

**Actual behavior**

**References**

**Additional context**

Discord username (if different from GitHub):donbikes

### LINKED ISSUE #1316: [BUG] ZFW/ZFWCG Data Entry Format
**Mod Version**
master (latest downloaded version)
built: 2020-10-14T04:45:18+00:00
ref: refs/heads/master
sha: 5ac2b03ea8132e7f549c3210e33437feef91f453
actor: Benjozork
event_name: push

**Describe the bug**
The ZFW entry allows entries that should be out of bounds. According to Appendix B of the Honeywell Pegasus Pilot Guide, the acceptable range of entries is 35.0-80.0 Metric Tonnes (or 77.1-176.0 !000s of pounds). I was able to enter "2" and also "900000."  For out-of bounds entries, there should be a "FORMAT ERROR" message in the scratchpad (see Appendix C of the Honeywell Pilot Guide).

Similarly, there should be some bounds on the ZFWCG entry of 8.0 to 50.0. I was unable to check this because once I put an entry in (and it was within the allowable range), it would not allow me to change that entry. This is also incorrect behavior. According to the Honeywell Pilots Guide (page 4-23), after initial entry, the ZFW and ZFWCG values cannot be cleared, but they can be modified either separately or together.

**To Reproduce**

1. Start Cold and Dark
2. Establish electrical power
3. Go to INIT B page of MCDU. Enter ZFW/ZFWCG using numbers that should be out of range. 
4. Once ZFW and ZFWCG entries have been accepted, try to change just the ZFWCG value.

**Expected behavior**
As stated above, numbers outside the allowable range should not be accepted and a FORMAT ERROR message should be displayed on the scratchpad.

For ZFWCG, modification of the value should be permitted.

**Actual behavior**
An out of range value for ZFW was accepted. I was unable to test out of range values for ZFWCG. Modification of the ZFWCG was not taken and should have been.

**References**

**Additional context**

Was this working before/when did the issue start occurring?Just happened to check it with this version when I was verifying that entries could even be made after Discord messages that 2 people were unable to enter ZFW/ZFWCG at all. 
Screenshot (32)
Screenshot (33)
Screenshot (31)

Is this a problem in the vanilla unmodded game?I don't know, but since the ZFW and ZFWCG entry process has been modified by the mod, I assume this issue is also fixable within the mod.

Discord username (if different from GitHub):donbikes#4084

### PR DESCRIPTION
**Pull request by Lucky38i**
Fixes #1316 
Fixes #219
Fixes #726
Fixes #689

## Summary of Changes

This will be ready to pull once #1652 is merged and I get dynamic updates working

- [x] Functional Init B inc. editable values and error handling
- [x] Functional Fuel Planning LSK in Init B
- [x] Automatic Change to FUEL PRED when engine is on
    - Fixed in #1724 
- [x] Convert RSV Fuel to EXTRA fuel on TO
- [x] Functional FUEL PRED inc editable values and error handling
  - [x]  Won't dynamically update until ##1772 is merged

Optional change
- [x] Simulate small delay on INIT B calculation, someone DM on discord if they have an idea.
    - Will be implemented when #1772 is merged otherwise will be done in a new PR

You can go ahead and test areas that are marked as completed and suggest any bugfixes

## For Developers
- There's a fuel prediction class containing several methods you may or may not find useful. Each one has JSDoc that should explain they're purpose and usage.
- I added math.js as I need some way to deal with matrices in a clean way.
- FMCMainDisplay has a couple for methods and attributes for values that weren't previously computed.
- DO NOT TOUCH THE COEFFECIENT VALUES

## For QA Testers
- INIT B simulates calculation delays that vary between 200ms to varying values based on route complexity, be wary of this if nothing immediately happens
- Test value boundaries on INIT B and FUEL PRED (reference honeywell manual)
- Test editable values (values in blue) and take note of fuel changes and font size changes
- attempt to enter values into green fields, you shouldn't be able to
- try using CLR and ensuring values revert back to predicted numbers

**Note**
- in order to use fuel planning you need to enter in a final time
- trips below 200NM and/or below FL290 produce experimental results as they're using alternate coefficients

**Currently not working**
- Extra Time
- computing FOB based on different sensors not currently modelled, therefore entering values in FOB won't do anything

## Screenshots (if necessary)

image
image
image
image
image
image

## References

- Honeywell Pegasus Step 1A (Rev 0), 2009 Section 4 Page 93
- Honeywell Pegasus Step 1A (Rev 0), 2009 Section 4 Page 22
- https://github.com/legoboyvdlp/A320-family/blob/dev/Nasal/FMGC/FMGC.nas
- the wonderful lads in #pilot-feedback
## Additional context

Discord username (if different from GitHub): Lucky38#3550

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created and uploaded.
The build.py will have already been run with the latest changes, so no need to rerun it once you download the zip.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the right side, slick on the **Artifacts** drop down and click the **A32NX** link

### CODE REVIEW

**Comment by MisterChocker** [comment]
> * Simulate small delay on INIT B calculation, someone DM on discord if they have an idea.

Do you have an option to do async call? this may to the job then

**Comment by Lucky38i** [comment]
> > * Simulate small delay on INIT B calculation, someone DM on discord if they have an idea.
> 
> Do you have an option to do async call? this may do the job then

Perhaps, haven't really dabbled with JS Async calls.

**Comment by MisterChocker** [comment]
> * Simulate small delay on INIT B calculation, someone DM on discord if they have an idea.

Now you know right?

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/FMC/A32NX_FMCMainDisplay.js [hunk 1]
```diff
@@ -707,7 +707,7 @@ class FMCMainDisplay extends BaseAirliners {
 
         let altToUse = this.cruiseFlightLevel;
         // Use the cruise level for calculations otherwise after cruise use descent altitude down to 10,000 feet.
-        if ([FlightPhase.FLIGHT_PHASE_DESCENT, FlightPhase.FLIGHT_PHASE_APPROACH, FlightPhase.FLIGHT_PHASE_GOAROUND].indexOf(this.currentFlightPhase) + 1) {
+        if ([FlightPhase.FLIGHT_PHASE_DESCENT, FlightPhase.FLIGHT_PHASE_APPROACH, FlightPhase.FLIGHT_PHASE_GOAROUND].includes(this.currentFlightPhase)) {
```
Comment (MisterChocker) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/FMC/A32NX_FMCMainDisplay.js:710]: `this.currentFlightPhase > FlightPhases.FLIGHT_PHASE_CRUISE`

**Review by MisterChocker — CHANGES_REQUESTED** [review]

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/FMC/A32NX_FMCMainDisplay.js (hunk 1 above)
Comment (Lucky38i) [reply A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/FMC/A32NX_FMCMainDisplay.js]: Thanks, still learning Enums lol. Fixed in 704dc35

**Comment by Benjozork** [comment]
LGTM

**Review by Benjozork — APPROVED** [review]

**Review by Carter633 — APPROVED** [review]
Name: British Avgeek UTube -- Carter633 #1436
Date of testing:  11/11/2020
Version of the sim: 1.10.11.0
PR Tested: #1524 
Object of test: Fuel Pred and Init B pages

Testing technique: Full Flight on VATSIM - EGKK to EGPH
Steps you've made to test the Pull Request:
1. Spawned Cold and Dark - used simbrief to generate route, inserted into MCDU
2. Added fuel and checked fuel pred and init B calculations - roughly matched simbrief
3. Departure, monitored fuel pred page - init B became "Not Available"
4. Had a large shortcut (10mins off flight time) by ATC enroute, checked fuel pred page and it adjusted dynamically
5. Checked FOB when parked at gate against what was predicted during flight - was basically bang on!

Flight notes
Notes written above, all good - no issues found
Route: LAM2X LAM L10 BPK/N0455F340 UN601 INPIP INPIP1E

After departure, given Direct LAM by ATC (cut all of the star out), then Direct SHAPP. At TARTN I flew one circuit in the hold as well to see if it changed fuel prediction again dynamically, it did.

Issues (If Any): None

Overall Rating: Good
Conclusions of the testing: Fuel Prediction page worked like a dream, Init B calculations all seemed correct too. Really great work!
Microsoft Flight Simulator 11_11_2020 13_15_22 (2)
Microsoft Flight Simulator 11_11_2020 13_08_37 (2)

**Comment by danbw9** [comment]
Name: danbw9#17028
Date of testing: 11/11/2020
Version of the sim: 1.10.11.0
PR Tested: #1524
Object of test: Testing out the FUEL PRED capability and INIT B

Testing technique: Try out on ground and then restart flight for full normal flight from UAAA to OMDB
Steps you've made to test the Pull Request: Check the items included in PR for QA testing above - and do a normal flight comparing versus a flight plan for fuel management. 

Spawn C&D on ground and put in all data in INIT A and check INIT B. 

I set my DEST ALT to OMSJ and initially after inputting the data in INIT B, it first gave me alternate fuel to -0.1 which was unusual. This I could correct to the proper value of 0.6 and all the values including EXTRA changed accordingly. 

Doing the MCDU setup in correct order when came to INIT B and the fuel data calculations were too slow. Though Lucky will change this. 

Cannot edit the green values. Checked I am able to input values in the blue fields and the font correctly increases to indicate pilot input. 

After putting in the average headwind component for the flight (Head 44 kts) it gave me a trip time of 0434 when actual flight was 0418, so not too far off. 

Changed taxi fuel from 0.2 to 0.3 and extra fuel value correctly decreased by 0.1.

On Takeoff, checked that the RTE RSV fuel was added to the EXTRA on takeoff which is all correct. 

Also confirmed that when I pretend as if my ALT Fuel is more than is required, that the MIN DEST FOB increases as well. 

Issues (If Any): Just a few:
- input lag little too slow
- fuel calculations for whole INIT B on ground should be quicker
- ALT FUEL giving -0.1 (will try again on other flights)

Overall Rating: Good
Conclusions of the testing: LGTM. I spoke with @Lucky38i on discord to clear some things up and give recommendations. The input delay lag for inputting data is a little too slow. As I have mentioned to @Lucky38i FINAL Reserve fuel is for 0030 mins and not 0045 - A320 being a jet aircraft.  It is changeable as the value is in blue so not an issue. Thank you for this nice addition, makes fuel planning on ground and in flight a lot easier!

**Comment by Lucky38i** [comment]
> I set my DEST ALT to OMSJ and initially after inputting the data in INIT B, it first gave me alternate fuel to -0.1 which was unusual. This I could correct to the proper value of 0.6 and all the values including EXTRA changed accordingly.

hey @danbw9 thanks for the review. The above is caused because your alternate is less than 20NM away from the destination (OMDB - OMSJ). My performance data can only handle down to 20NM and up to 500NM(This was the range of the Leap engine performance data I could find). I'm not sure how best to overcome this so it may be better suited that I just disable calculations outside of this range until I find more precise data. I may also use holding pattern traffic data to attempt to make predictions below 20NM.

This limitation is also shown on FUEL PRED when you're less than 20NM from the destination. the predictions just stop.

**Comment by Benjozork** [comment]
Hello, please rework to fit #1830

**Pull request merged**